### PR TITLE
KAFKA-2170: Updated Fixes For Windows Platform

### PR DESCRIPTION
**Pull request by GeorgeCGV**
**This is continuous development of [original pull request](https://github.com/apache/kafka/pull/3283) made by [nxmbriggs404](https://github.com/nxmbriggs404).**

I am using Kafka primary on Windows platform; but [KAFKA-1194](https://issues.apache.org/jira/browse/KAFKA-1194) affects all Kafka versions on Windows and makes it unusable without restarts. There are several pull requests available to fix this problem but most of them are outdated. [Nxmbriggs404 pull request](https://github.com/apache/kafka/pull/3283) is the most recent of them. Unfortunately, there is no further discussion and progress on forementioned pull request, so I created a new one.

Pull request contains the most recent merge of **apache/kafka** **trunk** branch into [original pull request](https://github.com/apache/kafka/pull/3283) made by [nxmbriggs404](https://github.com/nxmbriggs404) with all related tests and code part adaptations.

The only major change is in the `LogManager.scala` file within `asyncDelete`. I had to change operations order for `removedLog` from

1. Rename directory (atomic move) 
2. Make checkpoint log for offsets and recovery
3. Mark for deletion

to

1. Make checkpoint log for offsets and recovery
2. Close `FileChannel` to avoid `AccessDeniedException` on Windows
3. Rename directory (atomic move)
4. Mark for deletion

New order allows `LogManagerTest.testFileReferencesAfterAsyncDelete` to succeed on Windows.

Currently I am running long live test with all of the changes mentioned above. It seems to function as expected: log retention works. Setup is very simple - one broker.

TODO:
- `LogCleanerTest.testRecoveryAfterCrash` fails

There is potentially more broken parts on Windows... than the original bug

I think most Windows users would really like to see the old [KAFKA-1194](https://issues.apache.org/jira/browse/KAFKA-1194) issue being closed.

### CODE REVIEW

**Comment by jasonaliyetti** [comment]
@GeorgeCGV 
I'm looking at the same issue right now and I'm hitting an issue with the tests around transactions on Windows.  There seems to be some sort of bug in TransactionMarkerChannelManager that's causing this to hang in TransactionsTest.testBasicTransactions.  Have you run into this as well?

File: core/src/test/scala/unit/kafka/zk/EmbeddedZookeeper.scala [hunk 1]
```diff
@@ -59,6 +62,7 @@ class EmbeddedZookeeper() extends Logging {
 
     Iterator.continually(isDown()).exists(identity)
 
+    txnLog.close()
```
Comment (jasonaliyetti) [inline core/src/test/scala/unit/kafka/zk/EmbeddedZookeeper.scala:65]: I think you can just call `zookeeper.getTxnLogFactory.close()`

Comment (GeorgeCGV) [reply core/src/test/scala/unit/kafka/zk/EmbeddedZookeeper.scala:65]: It doesn't matter much; it will be the same obj. ref anyway

**Comment by GeorgeCGV** [comment]
@jasonaliyetti 
Indeed, transaction tests seems to be broken. 
Based on your previous comments in [PULL-3283](https://github.com/apache/kafka/pull/3283) it looks like they are not related to the [KAFKA-1194](https://issues.apache.org/jira/browse/KAFKA-1194) log rotation problem on Windows platform.

**Comment by jasonaliyetti** [comment]
They are separate issues but the end result is an unsuccessful build on Windows.

**Comment by MPeli** [comment]
Hi, I have created a new pull request. See #12331

**Pull request closed without merging**